extensions/poetry/main.ts

```diff
@@ -0,0 +1,53 @@
+import { PhylumApi } from "phylum"
+import { parse } from "https://deno.land/std/flags/mod.ts"
+
+// Add packages to the lockfile, and upload it to Phylum for analysis.
+// Return `true` if the packages pass the project's thresholds.
+// Return `false` if the packages don't pass the thresholds, or the processing
+// job is not finished.
+async function poetryCheck(args: string[]) {
+    console.log("Updating package lock…")
+
+    let process = Deno.run({ cmd: ["poetry", "add", '--lock', ...args] })
+    await process.status()
+    await process.close()
+
+    console.log("Package lock updated.\n")
+    console.log("Analyzing packages…")
+
+    const jobId = await PhylumApi.analyze("./poetry.lock")
+    const jobStatus = await PhylumApi.getJobStatus(jobId)
+
+    if (jobStatus.pass && jobStatus.status === "complete") {
+        console.log("All packages pass project thresholds.\n")
+        return true
+    } else if (jobStatus.pass) {
+        console.warn("Unknown packages were submitted for analysis, please check again later.")
+        return false
+    } else {
+        let packages = pkgs.join("', '")
```

Comment:
`pkgs` is not defined. Indeed, running this extension with a known bad package resulted in an error:

```
❯ /Users/maxrake/dev/phylum/localdev/cli/target/debug/phylum extension run /Users/maxrake/dev/phylum/localdev/cli/extensions/poetry add "pyyaml==5.3.1"
The `poetry` extension requires the following permissions:

  Read from the following paths:
    './poetry.lock'

  Run the following commands:
    'poetry'


Do you accept? yes
Updating package lock…

Updating dependencies
Resolving dependencies... (16.9s)

Writing lock file
Package lock updated.

Analyzing packages…
❗ Error: Execution failed caused by: ReferenceError: pkgs is not defined
    at poetryCheck (file:///Users/maxrake/dev/phylum/localdev/cli/extensions/poetry/main.ts:29:24)
    at async file:///Users/maxrake/dev/phylum/localdev/cli/extensions/poetry/main.ts:39:29
```

An unexpected side effect of this error was that the `pyproject.toml` and `poetry.lock` files were updated anyway. If this "failure" path is reached, it is expected that the bad package(s) would not be allowed and the files would not be updated.

Reply:
Good catch -- this is a leftover from previous commits where the arguments were parsed and the list of packages was extracted.

I would like to find an idiomatic way of rolling back a lockfile.
I'm no expert on poetry; can we rely on the following assumptions?
- `poetry add --lock` only updates `pyproject.toml` and `poetry.lock`
- `poetry add --lock` is not influenced by any files other than the above two

If that's the case, we could just simply copy those files in a temporary directory, run `poetry` in there, upload the updated lockfile, and then run `poetry add` again afterwards. This feels a bit wasteful because we run `poetry` twice, but I'm not sure whether there are better paths. Any ideas?

Reply:
> I would like to find an idiomatic way of rolling back a lockfile. I'm no expert on poetry; can we rely on the following assumptions?
> 
> * `poetry add --lock` only updates `pyproject.toml` and `poetry.lock`
> * `poetry add --lock` is not influenced by any files other than the above two

Yes, the belief is that these assumptions can be relied upon.

For other readers, this was discussed a bit offline, during an integrations team sync, and the idea of using temporary lock files (as it were) seemed to win the day. The original files would be "locked" while the Phylum extension runs and then either removed or atomically returned to their previous state, depending on the outcome and any error conditions of the extension logic.